### PR TITLE
fix(pipelined): Sessiond was deleted the enforcement and enforcement stat flows for 5G

### PR DESCRIPTION
**Pull request by prabinakpattnaik**
Signed-off-by: prabina pattnaik <prabinak@wavelabs.ai>

## Summary

Pipelined is sending the report message to sessiond about flow stats. If sessiond local info and pipelined stats info mismatch happening then sessiond has initiated the deactivate the mismatch flows. The sessiond local info is not consider the 5G flows, Hence table=13 and 14 not getting populated for 5G call flow.

Please find the logs:
Sep  2 07:57:55 magma-dev-focal sessiond[1824736]: I0902 07:57:55.492507 1824736 LocalEnforcer.cpp:392] Could not find an active session for IMSI222456000000001 and 192.168.128.254 during record aggregation
Sep  2 07:57:55 magma-dev-focal sessiond[1824736]: I0902 07:57:55.493772 1824736 LocalEnforcer.cpp:413] Received stats for 0 active sessions and 1 stale sessions
Sep  2 07:57:55 magma-dev-focal sessiond[1824736]: I0902 07:57:55.494238 1824736 LocalEnforcer.cpp:434] Deactivating flows for 1 dangling sessions in PipelineD
  
Sep  2 08:23:18 magma-dev-focal sessiond[1836888]: I0902 08:23:18.090224 1836888 GrpcMagmaUtils.cpp:58]
Sep  2 08:23:18 magma-dev-focal sessiond[1836888]:   magma.lte.DeactivateFlowsRequest {
Sep  2 08:23:18 magma-dev-focal sessiond[1836888]:       sid {
Sep  2 08:23:18 magma-dev-focal sessiond[1836888]:         id: "IMSI222456000000001"
Sep  2 08:23:18 magma-dev-focal sessiond[1836888]:       }
Sep  2 08:23:18 magma-dev-focal sessiond[1836888]:       request_origin {
Sep  2 08:23:18 magma-dev-focal sessiond[1836888]:         type: WILDCARD
Sep  2 08:23:18 magma-dev-focal sessiond[1836888]:       }
Sep  2 08:23:18 magma-dev-focal sessiond[1836888]:       ip_addr: "192.168.128.254"
Sep  2 08:23:18 magma-dev-focal sessiond[1836888]:       remove_default_drop_flows: true
Sep  2 08:23:18 magma-dev-focal sessiond[1836888]:       uplink_tunnel: 2147483647
Sep  2 08:23:18 magma-dev-focal sessiond[1836888]:   }

Solution
========
Extracted 5G flows from periodically stats report which is sending towards sessiond.

### CODE REVIEW

File: lte/gateway/python/magma/pipelined/app/enforcement_stats.py [hunk 1]
```diff
@@ -576,6 +576,8 @@ def _get_usage_from_flow_stat(self, flow_stats):
             ipv6_addr = _get_ipv6(flow_stat)
 
             local_f_teid_ng = _get_ng_local_f_id(flow_stat)
+            if local_f_teid_ng and report_flag:
```
Comment (ulaskozat) [inline lte/gateway/python/magma/pipelined/app/enforcement_stats.py:579]: this looks too hacky to me. I see an IGNORE_STATS facility used on line 570, specifying this in the policy itself makes more sense. @koolzz @pshelar what do you think?

Comment (ulaskozat) [reply lte/gateway/python/magma/pipelined/app/enforcement_stats.py:579]: second best option is to remove report_flag and put a comment over the check `if local_f_teid_nf` stating that when stat reporting feature is added for 5G, this check should be removed.

Comment (ulaskozat) [reply lte/gateway/python/magma/pipelined/app/enforcement_stats.py:579]: ideally, we should have used NO_TRACKING for this, but I see that in implementation it is not used inline with its meaning.

Comment (prabinakpattnaik) [reply lte/gateway/python/magma/pipelined/app/enforcement_stats.py:579]: Yes, it's a temp fix. We will add stat reporting feature for 5G on coming week then will remove the check.  The _get_usage_from_flow_stat is also used for delete_old_flows. So if report_flag is false means adding 5G flows as well for deleting old stats flows.

**Review by koolzz — CHANGES_REQUESTED** [review]
I'm confused whats happening here tbh. Why are we ignoring stats from get_stats? This function shoudn't even be called its not enabled by default yet(there is a sessiond flag for pull stats mode)

File: lte/gateway/python/magma/pipelined/app/enforcement_stats.py (hunk 1 above)
Comment (koolzz) [reply lte/gateway/python/magma/pipelined/app/enforcement_stats.py:579]: Ignore stats is internal implementation to allow few packets to egress prior to DPI classification(only enabled for subs with DPI based rules), but overall I'm not sure whats happening in this pr

**Comment by prabinakpattnaik** [comment]
> I'm confused whats happening here tbh. Why are we ignoring stats from get_stats? This function shoudn't even be called its not enabled by default yet(there is a sessiond flag for pull stats mode)

@koolzz , looks like EventOFPFlowStatsReply ofp event received by pipelined for every 2 seconds interval because of FlowStatsRequest requested by startup_flows.
def _poll_all_tables(self, datapath):
        """
        Send a FlowStatsRequest message to the datapath
        """
        ofproto, parser = datapath.ofproto, datapath.ofproto_parser
        req = parser.OFPFlowStatsRequest(
            datapath,
            out_group=ofproto.OFPG_ANY,
            out_port=ofproto.OFPP_ANY,
        )
Sep  4 19:08:06 magma-dev-focal pipelined[11374]: INFO:root:ofp_event.EventOFPFlowStatsReply
Sep  4 19:08:06 magma-dev-focal pipelined[11374]: INFO:EnforcementStatsController:Sending RPC payload: RuleRecordTable{
Sep  4 19:08:06 magma-dev-focal pipelined[11374]: epoch: 1630782470
Sep  4 19:08:06 magma-dev-focal pipelined[11374]: }
Sep  4 19:08:08 magma-dev-focal pipelined[11374]: INFO:root:ofp_event.EventOFPFlowStatsReply
Sep  4 19:08:08 magma-dev-focal pipelined[11374]: INFO:EnforcementStatsController:Sending RPC payload: RuleRecordTable{
Sep  4 19:08:08 magma-dev-focal pipelined[11374]: epoch: 1630782470
Sep  4 19:08:08 magma-dev-focal pipelined[11374]: }
Sep  4 19:08:13 magma-dev-focal pipelined[11374]: INFO:root:ofp_event.EventOFPFlowStatsReply
Sep  4 19:08:13 magma-dev-focal pipelined[11374]: INFO:EnforcementStatsController:Sending RPC payload: RuleRecordTable{
Sep  4 19:08:13 magma-dev-focal pipelined[11374]: epoch: 1630782470
Sep  4 19:08:13 magma-dev-focal pipelined[11374]: }

File: lte/gateway/python/magma/pipelined/app/enforcement_stats.py (hunk 1 above)
Comment (koolzz) [reply lte/gateway/python/magma/pipelined/app/enforcement_stats.py:579]: Just leave if local_teid and completely remove the  report flag and I'll accept

**Review by koolzz — CHANGES_REQUESTED** [review]

**Comment by ulaskozat** [comment]
is there a reason why this PR is still open?

**Comment by ulaskozat** [comment]
is there a reason why this PR is still open?

**Comment by prabinakpattnaik** [comment]
> is there a reason why this PR is still open?

Sorry, not required for this PR. We have raised the PR for usage reporting for 5G

**Pull request closed without merging**